hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/source/enumerator/HoodieContinuousSplitEnumerator.java

```diff
@@ -0,0 +1,134 @@
+ */
+public class HoodieContinuousSplitEnumerator extends AbstractHoodieSplitEnumerator {
+  private static final Logger LOG = LoggerFactory.getLogger(HoodieContinuousSplitEnumerator.class);
+
+  private final SplitEnumeratorContext<HoodieSourceSplit> enumeratorContext;
+  private final HoodieSplitProvider splitProvider;
+  private final HoodieContinuousSplitDiscover splitDiscover;
+  private final ScanContext scanContext;
+
+  /**
+   * Instant for the last enumerated commit. Next incremental enumeration should be based off
+   * this as the starting position.
+   */
+  private final AtomicReference<HoodieEnumeratorPosition> position;
+
+  public HoodieContinuousSplitEnumerator(
+      SplitEnumeratorContext<HoodieSourceSplit> enumeratorContext,
+      HoodieSplitProvider splitProvider,
+      HoodieContinuousSplitDiscover splitDiscover,
+      ScanContext scanContext,
+      @Nullable HoodieSplitEnumeratorState enumState) {
+    super(enumeratorContext, splitProvider);
+    this.enumeratorContext = enumeratorContext;
+    this.splitProvider = splitProvider;
+    this.splitDiscover = splitDiscover;
+    this.scanContext = scanContext;
+    this.position = new AtomicReference<>();
+
+    if (enumState != null) {
+      this.position.set(HoodieEnumeratorPosition.of(enumState.getLastEnumeratedInstant(), enumState.getLastEnumeratedInstantCompletionTime()));
+    } else {
+      this.position.set(HoodieEnumeratorPosition.of(scanContext.getStartInstant(), scanContext.getStartInstant()));
+    }
+  }
+
+  @Override
+  public void start() {
+    super.start();
+    enumeratorContext.callAsync(
+        this::discoverSplits,
+        this::processDiscoveredSplits,
+        0L,
+        scanContext.getScanInterval().toMillis());
+  }
+
+  @Override
+  boolean shouldWaitForMoreSplits() {
+    return true;
+  }
+
+  @Override
+  public HoodieSplitEnumeratorState snapshotState(long checkpointId) throws Exception {
+    return new HoodieSplitEnumeratorState(splitProvider.state(), position.get().lastInstant(), position.get().lastInstantCompletionTime());
+  }
+
+  private HoodieContinuousSplitBatch discoverSplits() {
+    int pendingSplitNumber = splitProvider.pendingSplitCount();
+    if (pendingSplitNumber > scanContext.getMaxPendingSplits()) {
+      LOG.info(
+          "Pause split discovery as the assigner already has too many pending splits: {}",
+          pendingSplitNumber);
+      return HoodieContinuousSplitBatch.EMPTY;
+    }
+    return splitDiscover.discoverSplits(position.get().lastInstantCompletionTime());
+  }
+
+  private void processDiscoveredSplits(HoodieContinuousSplitBatch result, Throwable throwable) {
+    if (throwable == null) {
+      if (!Objects.equals(result.getFromInstant(), position.get().lastInstantCompletionTime())) {
+        LOG.info(
+            "Skip {} discovered splits because the scan starting position doesn't match "
```

Comment:
Curious about when will this happen, I notice the `fromInstant` of `result` is just assigned as `position.get().lastInstantCompletionTime()` in `discoverSplits`, so this checking will never happen?

Reply:
It is just for protection.